src/python/grpcio/grpc/experimental/aio/_channel.py

```diff
@@ -383,22 +351,32 @@ def __init__(self, target: str, options: ChannelArgumentType,
         # No new calls will be accepted by the Cython channel.
         self._channel.closing()
 
-        if grace:
-            # pylint: disable=unused-variable
-            _, pending = await asyncio.wait(self._ongoing_calls.calls,
-                                            timeout=grace,
-                                            loop=self._loop)
-
-            if not pending:
-                return
-
-        # A new set is created acting as a shallow copy because
-        # when cancellation happens the calls are automatically
-        # removed from the originally set.
-        calls = WeakSet(data=self._ongoing_calls.calls)
+        # Iterate through running tasks
+        tasks = _all_tasks()
+        calls = []
+        call_tasks = []
+        for task in tasks:
+            stack = task.get_stack(limit=1)
+            if not stack:
+                continue
+
+            # Locate ones created by `aio.Call`.
+            frame = stack[0]
+            if 'self' in frame.f_locals:
```

Comment:
We're doing several lookups in this map for the same element. Can we get away with just a single `frame.f_locals.get('self')`?

Reply:
or catch the attribute lookup
```
f_locals = frame.f_locals
```

Reply:
Good idea for the optimization. Updated to fetch the value only once.